common/components/diagonal_block_manipulation.hpp.inc

```diff
@@ -58,38 +58,36 @@ __device__ __forceinline__ void extract_transposed_diag_blocks(
     IndexType bsize = 0;
 #pragma unroll
     for (int b = 0; b < processed_blocks; ++b, ++bid) {
-        if (bid >= num_blocks) {
-            break;
-        }
-        bstart += bsize;
-        bsize = block_ptrs[bid + 1] - bstart;
+        if (bid < num_blocks) {
+            bstart += bsize;
+            bsize = block_ptrs[bid + 1] - bstart;
 #pragma unroll
-        for (int i = 0; i < max_block_size; ++i) {
-            if (i >= bsize) {
-                break;
-            }
-            if (threadIdx.y == b && threadIdx.x < max_block_size) {
-                workspace[threadIdx.x] = zero<ValueType>();
-            }
-            warp.sync();
-            const auto row = bstart + i;
-            const auto rstart = row_ptrs[row] + tid;
-            const auto rend = row_ptrs[row + 1];
-            // use the entire warp to ensure coalesced memory access
-            for (auto j = rstart; j < rend; j += config::warp_size) {
-                const auto col = col_idxs[j] - bstart;
-                if (col >= bsize) {
-                    break;
+            for (int i = 0; i < max_block_size; ++i) {
+                if (i < bsize) {
+                    if (threadIdx.y == b && threadIdx.x < max_block_size) {
+                        workspace[threadIdx.x] = zero<ValueType>();
+                    }
+                    warp.sync();
+                    const auto row = bstart + i;
+                    const auto rstart = row_ptrs[row] + tid;
+                    const auto rend = row_ptrs[row + 1];
+                    // use the entire warp to ensure coalesced memory access
+                    for (auto j = rstart; j < rend; j += config::warp_size) {
```

Comment:
the `rend` is not determined in compile time, so I think it can not be unroll.

Reply:
My point is: These unrolls seem unnecessary at best, and counterproductive at worst, since usually the point of unrolling is to optimize a small, hot loop for throughput, but here the loops we unroll (at lines 65 and 60) cannot profit from unrolling, since the actual work happens in a variable-sized (even with different per-thread trip counts) loop that always needs some kind of branching/jumps to be executed.

Reply:
You are right. Unrolling should be on the small and simple loop.

Reply:
After looking again into #289 again, I actually put `pragma unroll 1` only for some of the unroll in the file `common/components/warp_blas.hpp.inc` (I thought I had done that everywhere for the jacobi generate), which is where the critical parts seemed to be coming from at the time.  So it might be possible that this one unroll is overkill and doesn't give anything extra, it would need to be benchmarked (probably also with assembly comparison).